File: uikit/src/main/java/com/kotlin/in_app_notification/InAppNotification.java
package com.kotlin.in_app_notification;

import android.content.Context;
import android.content.res.Resources;
import android.graphics.PorterDuff;
import android.graphics.drawable.StateListDrawable;
import android.support.annotation.ColorRes;
import android.support.annotation.Dimension;
import android.support.annotation.DrawableRes;
import android.support.annotation.NonNull;
import android.support.design.widget.Snackbar;
import android.support.v4.graphics.ColorUtils;
import android.util.DisplayMetrics;
import android.util.TypedValue;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import com.bumptech.glide.Glide;
import com.example.uikit.R;
import de.hdodenhof.circleimageview.CircleImageView;

public final class InAppNotification {
    private Snackbar mSnackbar;
    private View custom;
    private Resources resources;

    public static InAppNotification make(
            View parent,
            String title,
            String text,
            @ColorRes int backgroundColorInt,
            @DrawableRes int imageResId
    ){
        InAppNotification notification = new InAppNotification(parent, title, text, backgroundColorInt);
        notification.setImage(notification.custom, imageResId);
        return notification;
    }

    public static InAppNotification make(
            View parent,
            String title,
            String text,
            @ColorRes int backgroundColorInt,
            String imageURL
    ){
        InAppNotification notification = new InAppNotification(parent, title, text, backgroundColorInt);
        notification.setImage(notification.custom, imageURL);
        return notification;
    }

    private InAppNotification(
            View parent,
            String title,
            String text,
            @ColorRes int backgroundColorInt
    ){
        this.mSnackbar = Snackbar.make(parent, "", Snackbar.LENGTH_SHORT);
        this.resources = parent.getResources();

        Context context = parent.getContext();

        this.custom = LayoutInflater.from(context).inflate(R.layout.in_app_notification, null);
        ((ViewGroup) mSnackbar.getView()).removeAllViews();
        ((ViewGroup) mSnackbar.getView()).addView(custom);


        TextView titleTV = custom.findViewById(R.id.notificationTitle);
        titleTV.setText(title);
        TextView textTV = custom.findViewById(R.id.notificationText);
        textTV.setText(text);

        CircleImageView close = custom.findViewById(R.id.notificationClose);
        close.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                mSnackbar.dismiss();
            }
        });

        ViewGroup notification = custom.findViewById(R.id.in_app_notification);
        StateListDrawable shapeDrawable = (StateListDrawable) notification.getBackground();
        int colorBackground = resources.getColor(backgroundColorInt, null);
        shapeDrawable.setColorFilter(colorBackground, PorterDuff.Mode.MULTIPLY);

        if (isColorDark(colorBackground)) {
            titleTV.setTextColor(resources.getColor(R.color.absolute_white, null));
            textTV.setTextColor(resources.getColor(R.color.absolute_white, null));
            close.setColorFilter(resources.getColor(R.color.absolute_white, null));
            close.setBorderColor(resources.getColor(R.color.absolute_white, null));
        }
        configSnackDesign();
    }

    private boolean isColorDark(int color) {
        return ColorUtils.calculateLuminance(color) < 0.7;
    }

    private void configSnackDesign() {
        ViewGroup.MarginLayoutParams params = (ViewGroup.MarginLayoutParams) mSnackbar.getView().getLayoutParams();
        params.setMargins(dpToPx(16), dpToPx(0), dpToPx(16), dpToPx(90));

        mSnackbar.getView().setLayoutParams(params);
        mSnackbar.getView().setBackgroundResource(R.drawable.foreground_notification);
        mSnackbar.getView().setPadding(dpToPx(0), dpToPx(0), dpToPx(0), dpToPx(0));
        mSnackbar.getView().setElevation(10);
    }

    private int dpToPx(@Dimension(unit = Dimension.DP) int dip) {
        final DisplayMetrics displayMetrics = resources.getDisplayMetrics();
        return (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, dip, displayMetrics);
    }

    private void setImage(@NonNull View custom, @DrawableRes int drawableImageID){
        CircleImageView image = custom.findViewById(R.id.notificationImage);
        image.setImageResource(drawableImageID);
    }

    private void setImage(@NonNull View custom, String imageURL) {
        CircleImageView image = custom.findViewById(R.id.notificationImage);
        Glide.with(custom)
                .load(imageURL)
                .into(image);
    }

    public void show(){
        mSnackbar.show();
    }

    public void setIndefinite(boolean isIndefinite) {
        if (isIndefinite) {
            mSnackbar.setDuration(Snackbar.LENGTH_INDEFINITE);
        } else {
            mSnackbar.setDuration(Snackbar.LENGTH_SHORT);
        }
    }
}

File: uikit/src/main/java/com/kotlin/in_app_notification/InAppNotification.kt
package com.kotlin.in_app_notification

import android.content.res.Resources
import android.graphics.PorterDuff
import android.graphics.drawable.StateListDrawable
import android.support.annotation.ColorRes
import android.support.annotation.Dimension
import android.support.annotation.DrawableRes
import android.support.design.widget.Snackbar
import android.support.v4.graphics.ColorUtils
import android.util.TypedValue
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.bumptech.glide.Glide
import com.example.uikit.R
import de.hdodenhof.circleimageview.CircleImageView

class InAppNotification private constructor(
        parent: View,
        title: String,
        text: String,
        @ColorRes backgroundColorInt: Int
) {
    var onNotificationClickListener: OnNotificationClickListener? = null
    private val mSnackbar: Snackbar = Snackbar.make(parent, "", Snackbar.LENGTH_SHORT)
    private val custom: View = LayoutInflater.from(parent.context).inflate(R.layout.in_app_notification, null)
    private val resources: Resources = parent.resources

    init {
        (mSnackbar.view as ViewGroup).removeAllViews()
        (mSnackbar.view as ViewGroup).addView(custom)

        val titleTV = custom.findViewById<TextView>(R.id.notificationTitle)
        titleTV.text = title
        val textTV = custom.findViewById<TextView>(R.id.notificationText)
        textTV.text = text

        val close = custom.findViewById<CircleImageView>(R.id.notificationClose)
        close.setOnClickListener { mSnackbar.dismiss() }

        val notification = custom.findViewById<ViewGroup>(R.id.in_app_notification)
        val shapeDrawable = notification.background as StateListDrawable
        val colorBackground = resources.getColor(backgroundColorInt, null)
        shapeDrawable.setColorFilter(colorBackground, PorterDuff.Mode.MULTIPLY)

        if (isColorDark(colorBackground)) {
            titleTV.setTextColor(resources.getColor(R.color.absolute_white, null))
            textTV.setTextColor(resources.getColor(R.color.absolute_white, null))
            close.setColorFilter(resources.getColor(R.color.absolute_white, null))
            close.borderColor = resources.getColor(R.color.absolute_white, null)
        }
        configSnackDesign()

        custom.setOnClickListener(View.OnClickListener {
            onNotificationClickListener?.let {
                it.onNotificationClick()
            }
        })
    }

    private fun isColorDark(color: Int): Boolean {
        return ColorUtils.calculateLuminance(color) < 0.7
    }

    private fun configSnackDesign() {
        val params = mSnackbar.view.layoutParams as ViewGroup.MarginLayoutParams
        params.setMargins(dpToPx(16), dpToPx(0), dpToPx(16), dpToPx(90))

        mSnackbar.view.layoutParams = params
        mSnackbar.view.setBackgroundResource(R.drawable.foreground_notification)
        mSnackbar.view.setPadding(dpToPx(0), dpToPx(0), dpToPx(0), dpToPx(0))
        mSnackbar.view.elevation = 10f
    }

    private fun dpToPx(@Dimension(unit = Dimension.DP) dip: Int): Int {
        val displayMetrics = resources.displayMetrics
        return TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, dip.toFloat(), displayMetrics).toInt()
    }

    private fun setImage(custom: View, @DrawableRes drawableImageID: Int) {
        val image = custom.findViewById<CircleImageView>(R.id.notificationImage)
        image.setImageResource(drawableImageID)
    }

    private fun setImage(custom: View, imageURL: String) {
        val image = custom.findViewById<CircleImageView>(R.id.notificationImage)
        Glide.with(custom)
                .load(imageURL)
                .into(image)
    }

    fun show() {
        mSnackbar.show()
    }

    fun setIndefinite(isIndefinite: Boolean) {
        mSnackbar.duration = if (isIndefinite) Snackbar.LENGTH_INDEFINITE else Snackbar.LENGTH_SHORT
    }

    companion object {

        fun make(
                parent: View,
                title: String,
                text: String,
                @ColorRes backgroundColorInt: Int,
                @DrawableRes imageResId: Int
        ): InAppNotification {
            val notification = InAppNotification(parent, title, text, backgroundColorInt)
            notification.setImage(notification.custom, imageResId)
            return notification
        }

        fun make(
                parent: View,
                title: String,
                text: String,
                @ColorRes backgroundColorInt: Int,
                imageURL: String
        ): InAppNotification {
            val notification = InAppNotification(parent, title, text, backgroundColorInt)
            notification.setImage(notification.custom, imageURL)
            return notification
        }
    }
}
